api/src/utils/transformations.ts

```diff
@@ -63,3 +113,72 @@ export function maybeExtractFormat(transforms: Transformation[]): string | undef
 	const lastToFormat = toFormats[toFormats.length - 1];
 	return lastToFormat ? lastToFormat[1]?.toString() : undefined;
 }
+
+type Dimensions = { w: number; h: number };
+type FocalPoint = { x: number; y: number };
+
+/**
+ * Resize an image but keep it centered on the focal point.
+ * Based on the method outlined in https://github.com/lovell/sharp/issues/1198#issuecomment-384591756
+ */
+function getResizeArguments(
+	original: Dimensions,
+	target: Dimensions,
+	focalPoint?: FocalPoint | null,
+): { width: number; height: number; region: Region } {
+	const { width, height, factor } = getIntermediateDimensions(original, target);
+
+	const region = getExtractionRegion(factor, focalPoint ?? { x: original.w / 2, y: original.h / 2 }, target, {
+		w: width,
+		h: height,
+	});
+
+	return { width, height, region };
+}
+
+/**
+ * Calculates the dimensions of the intermediate (resized) image.
+ */
+function getIntermediateDimensions(
+	original: Dimensions,
+	target: Dimensions,
+): { width: number; height: number; factor: number } {
+	const hRatio = original.h / target.h;
+	const wRatio = original.w / target.w;
+
+	let factor: number;
+	let width: number;
+	let height: number;
+
+	if (hRatio < wRatio) {
+		factor = hRatio;
+		height = Math.round(target.h);
+		width = Math.round(original.w / factor);
+	} else {
+		factor = wRatio;
+		width = Math.round(target.w);
+		height = Math.round(original.h / factor);
+	}
+
+	return { width, height, factor };
+}
+
+/**
+ * Calculates the Region to extract from the intermediate image.
+ */
+function getExtractionRegion(
+	factor: number,
+	focalPoint: FocalPoint,
+	target: Dimensions,
+	intermediate: Dimensions,
+): Region {
+	const newXCenter = focalPoint.x / factor;
+	const newYCenter = focalPoint.y / factor;
```

Comment:
Do we need to check if `factor` is not `0` to prevent NaN's in the calculations? I think `clamp` will filter it out but does that affect the calculation done here?